pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,407 @@
+	)
+	bfs.ctx, bfs.ctxCancel = context.WithCancel(context.Background())
+	if err = bfs.pool.Dial(context.Background()); err != nil {
+		bfs.isActive.CompareAndSwap(true, false)
+		return fmt.Errorf("failed to dial NeoFS pool: %w", err)
+	}
+
+	err = containerID.DecodeString(bfs.cfg.ContainerID)
+	if err != nil {
+		bfs.isActive.CompareAndSwap(true, false)
+		return fmt.Errorf("failed to decode container ID: %w", err)
+	}
+
+	err = bfs.retry(func() error {
+		containerObj, err = bfs.pool.ContainerGet(bfs.ctx, containerID, client.PrmContainerGet{})
+		return err
+	})
+	if err != nil {
+		bfs.isActive.CompareAndSwap(true, false)
+		return fmt.Errorf("failed to get container: %w", err)
+	}
+	containerMagic := containerObj.Attribute("Magic")
+	if containerMagic != strconv.Itoa(int(bfs.chain.GetConfig().Magic)) {
+		bfs.isActive.CompareAndSwap(true, false)
+		return fmt.Errorf("container magic mismatch: expected %d, got %s", bfs.chain.GetConfig().Magic, containerMagic)
+	}
+	currentHeight := bfs.chain.HeaderHeight()
+	var lastSavedRoot *state.MPTRoot
+	var lastSavedHeight uint32
+	for h := currentHeight; h > 0; h-- {
+		root, err := bfs.stateModule.GetStateRoot(h)
+		if err == nil {
+			lastSavedRoot = root
+			lastSavedHeight = h
+			break
+		}
+	}
+	if lastSavedRoot != nil {
+		bfs.log.Info("found last saved state root in store",
+			zap.String("root", lastSavedRoot.Root.String()),
+			zap.Uint32("height", lastSavedHeight))
+	} else {
+		bfs.log.Info("no saved state root found in store up to current height",
+			zap.Uint32("currentHeight", currentHeight))
+	}
+	if lastSavedHeight == currentHeight-1 {
+		bfs.log.Info("no new state root to fetch, exiting")
+		return nil
+	}
+	reader, err := bfs.getMPTNodes()
+	if err != nil {
+		bfs.log.Error("failed to get MPT nodes", zap.Error(err))
+		return err
+	}
+	defer reader.Close()
+
+	br := gio.NewBinReaderFromIO(reader)
+
+	version := br.ReadB()
+	if version != 0 || br.Err != nil {
+		return fmt.Errorf("invalid version: %d, error: %w", version, br.Err)
+	}
+	_ = br.ReadU32LE()
+	syncHeight := br.ReadU32LE()
+	var stateRoot util.Uint256
+	br.ReadBytes(stateRoot[:])
+	if br.Err != nil {
+		return fmt.Errorf("failed to read NeoFS header: %w", br.Err)
+	}
+	if syncHeight > currentHeight {
+		return fmt.Errorf("sync height %d exceeds current height %d", syncHeight, currentHeight)
+	}
+
+	bfs.log.Info("NeoFS state root", zap.String("root", stateRoot.String()), zap.Uint32("height", syncHeight))
+
+	localMPT := mpt.NewTrie(mpt.EmptyNode{}, mpt.ModeAll, bfs.stateModule.Store)
+	bfs.log.Info("initial MPT root", zap.String("root", localMPT.StateRoot().String()))
+	batch := make(map[string][]byte, batchSize)
+	processed := 0
+	cache := storage.NewMemCachedStore(bfs.stateModule.Store)
+	tempPrefix := []byte{byte(storage.STTempStorage)}
+
+	for {
+		neoFSKey := br.ReadVarBytes()
+		if br.Err != nil {
+			if errors.Is(br.Err, io.EOF) {
+				break
+			}
+			bfs.log.Error("failed to read key", zap.Error(br.Err))
```

Comment:
As a node's operator I'll see "failed to read key" error in the node log's message. I don't have a clue what is the source of this message. How am I supposed to understand where does it com from?

Check/adjust all relevant places.